app/Http/Controllers/AlbumController.php

```diff
@@ -541,123 +550,183 @@ public function getArchive(Request $request)
+						if ($this->sessionFunctions->is_logged_in()) {
+							$user = User::find($UserId);
+
+							if ($UserId == 0 || $user->upload) {
+								$photos_sql = Photo::select_stars(Photo::OwnedBy($UserId));
+								break;
+							}
+						}
+						$photos_sql = Photo::select_stars(Photo::whereIn('album_id', $this->albumFunctions->getPublicAlbums()));
+						break;
+					case 's':
+						$dir = 'Public';
+						$photos_sql = Photo::select_public(Photo::OwnedBy($UserId));
+						break;
+					case 'r':
+						$dir = 'Recent';
+						if ($this->sessionFunctions->is_logged_in()) {
+							$user = User::find($UserId);
+
+							if ($UserId == 0 || $user->upload) {
+								$photos_sql = Photo::select_recent(Photo::OwnedBy($UserId));
+								break;
+							}
+						}
+						$photos_sql = Photo::select_recent(Photo::whereIn('album_id', $this->albumFunctions->getPublicAlbums()));
+						break;
+					case '0':
+						$dir = 'Unsorted';
+						$photos_sql = Photo::select_unsorted(Photo::OwnedBy($UserId));
+						break;
+					default:
+						$album = Album::find($albumID);
+						if ($album === null) {
+							Logs::error(__METHOD__, __LINE__, 'Could not find specified album');
+
+							return 'false';
+						}
+						$dir = $album->title;
+						$photos_sql = Photo::set_order(Photo::where('album_id', '=', $albumID));
+						break;
+				} // switch (albumID)
 
-			$photos = $photos_sql->get();
-			foreach ($photos as $photo) {
-				$title = str_replace($badChars, '', $photo->title);
-				$url = Config::get('defines.urls.LYCHEE_URL_UPLOADS_BIG') . $photo->url;
+				$compress_album = function ($photos_sql, $dir, &$dirs, $parent_dir, $album) use (&$zip, $badChars, &$compress_album) {
+					$dir = str_replace($badChars, '', $dir);
+					if ($dir === '') {
+						$dir = 'Untitled';
+					}
+					// Check for duplicates
+					if (!empty($dirs)) {
+						$i = 1;
+						$tmp_dir = $dir;
+						while (in_array($tmp_dir, $dirs)) {
+							// Set new directory name
+							$tmp_dir = $dir . '-' . $i;
+							$i++;
+						}
+						$dir = $tmp_dir;
+					}
+					$dirs[] = $dir;
 
-				if (!isset($title) || $title === '') {
-					$title = 'Untitled';
-				}
-				// Check if readable
-				if (!@is_readable($url)) {
-					Logs::error(__METHOD__, __LINE__, 'Original photo missing: ' . $url);
-					continue;
-				}
+					if ($parent_dir !== '') {
+						$dir = $parent_dir . '/' . $dir;
+					}
 
-				// Get extension of image
-				$extension = Helpers::getExtension($url, false);
-				// Set title for photo
-				$zipFileName = $zipTitle . '/' . $title . $extension;
-				// Check for duplicates
-				if (!empty($files)) {
-					$i = 1;
-					while (in_array($zipFileName, $files)) {
-						// Set new title for photo
-						$zipFileName = $zipTitle . '/' . $title . '-' . $i . $extension;
-						$i++;
+					$files = [];
+					$photos = $photos_sql->get();
+					foreach ($photos as $photo) {
+						$url = Config::get('defines.urls.LYCHEE_URL_UPLOADS_BIG') . $photo->url;
```

Comment:
this will need to use the Storage facade (after merging the image protection one).